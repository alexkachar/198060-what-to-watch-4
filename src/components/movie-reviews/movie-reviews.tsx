import * as React from 'react';
import {connect} from 'react-redux';

import {getReviews} from '../../store/reducers/review/selectors';
import ReviewOperation from '../../store/operations/review/review';
import {checkIfEven, checkIfOdd} from '../../utils';

import ReviewCard from '../review-card/review-card';
import Review from '../../interfaces/review';
import Loader from '../loader/loader';

interface Props {
  movieId: number | string;
  reviews: Review[];
  onRequestReviews: (movieId: string | number) => void;
}

class MovieReviews extends React.PureComponent<Props> {

  componentDidMount() {
    const {onRequestReviews, movieId} = this.props;
    onRequestReviews(movieId);
  }

  componentDidUpdate(prevProps) {
    const {onRequestReviews, movieId} = this.props;

    if (movieId !== prevProps.movieId) {
      onRequestReviews(movieId);
    }
  }

  render() {

    const {reviews} = this.props;

    if (!reviews) {
      return <Loader />;
    }

    return (
      <div className="movie-card__reviews movie-card__row">

        <div className="movie-card__reviews-col">

          {reviews.map((review, i) => checkIfOdd(i + 1) ? <ReviewCard review={review} key={review.id} /> : null)}

        </div>

        <div className="movie-card__reviews-col">
          {reviews.map((review, i) => checkIfEven(i + 1) ? <ReviewCard review={review} key={review.id} /> : null)}
        </div>
      </div>
    );
  }

}

const mapStateToProps = (state) => ({
  reviews: getReviews(state),
});

const mapDispatchToProps = (dispatch) => ({

  onRequestReviews: (movieId) => {
    dispatch(ReviewOperation.loadReviews(movieId));
  }

});

export {MovieReviews};
export default connect(mapStateToProps, mapDispatchToProps)(MovieReviews);
