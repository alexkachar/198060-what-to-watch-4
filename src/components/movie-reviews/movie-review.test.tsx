import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {MovieReviews} from './movie-reviews';
import Review from '../../interfaces/review';
import mockReviews from '../../test-data/mock-reviews';

const REVIEWS: Review[] = mockReviews;

it(`MovieReviews renders correctly`, () => {
  const tree = renderer
  .create(
      <MovieReviews
        movieId={1}
        reviews={REVIEWS}
        onRequestReviews={jest.fn()}
      />, {
        createNodeMock: () => document.createElement(`div`)
      })
.toJSON();
  expect(tree).toMatchSnapshot();
});
