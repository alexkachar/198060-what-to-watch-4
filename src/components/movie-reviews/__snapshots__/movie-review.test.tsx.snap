// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieReviews renders correctly 1`] = `
<div
  className="movie-card__reviews movie-card__row"
>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Kate Muir
          </cite>
          <time
            className="review__date"
            dateTime="2019-05-08T14:13:56.569Z"
          >
            May 8, 2019
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        8.9
      </div>
    </div>
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          The mannered, madcap proceedings are often delightful, occasionally silly, and here and there, gruesome and/or heartbreaking.
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Amanda Greever
          </cite>
          <time
            className="review__date"
            dateTime="2020-06-02T13:12:46.569Z"
          >
            June 2, 2020
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        8.2
      </div>
    </div>
  </div>
</div>
`;
